### PR TITLE
memory cards game

### PR DESCRIPTION
**Pull request by semenjob**
[Demo](https://semenjob.github.io/pair-game/) |
[Code base](https://github.com/semenjob/pair-game)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

### CODE REVIEW

**Comment by OleksiyRudenko** [comment]
Being a fan of MPG I tried it.

It would be nice to see all cards without need to scroll. Would be great if you could apply learnings from past tasks to make UX better...

Also, at some point I clicked randomly and here is what has happened:

image

**Comment by OleksiyRudenko** [comment]
A lot better UX now 👍 

Disclaimer: this is not a code review feedback.

File: submissions/semenjob/memoryCards/game-logic.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  card.dataset.cardId = id;
+  return card;
+};
+
+const toggleFlip = (card) => {
+  if (card !== null) {
+    card.classList.toggle("flip");
+  }
+  return;
+};
```
Comment (dromix) [inline submissions/semenjob/memoryCards/game-logic.js:28]:
```suggestion
const toggleMemoryCard = (card) => card?.classList.toggle("flip");
```

File: submissions/semenjob/memoryCards/game-logic.js [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  });
+
+  function dubleCheckCards(secondCards) {
+    const isEqual = secondCards.dataset.cardId === firstCards.dataset.cardId;
+
+    if (!isEqual) {
+      setTimeout(() => {
+        toggleFlip(secondCards);
+        toggleFlip(firstCards);
+        firstCards = null;
```
Comment (dromix) [inline submissions/semenjob/memoryCards/game-logic.js:76]: I would suggest to create some structure for pairCards like
```javascript
class PairCards {
  firstCard;
  secondCard; // or even provide linked-list structure if memory card can be more than 2 pairs

  setFirstCards(card) {
    this.firstCard = card;
  }

  setSecondCard(card) {
    this.secondCard = card;
  }

  check() {} // validates that cards are equel
  reset() {} // toggle cards to init state
}
```
Usage example:
```javascript
const pair = new PairCards();
pair.setFirstCard(card);
pair.setSecondCard(card);

const isEquel = pair.check();
if (!isEquel) pair.reset();
```

**Comment by kasionio** [comment]
Seems it is not working well 
image

File: submissions/semenjob/memoryCards/game-logic.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+  for (let i = 0; i < cardIndexes.length; i++) {
+    const cardId = cardIndexes[i];
+    const backgroundImage = memoryCardImages[cardId];
+    const card = createMemoryCard(
+      cardId,
+      memoryCardFrontImage,
+      backgroundImage
+    );
+    cards.append(card);
```
Comment (A-Ostrovnyy) [inline submissions/semenjob/memoryCards/game-logic.js:55]: At this line, you append a card on each iteration, avoid that pattern. 
Use `append` method, which can add several elements in one operation

File: submissions/semenjob/memoryCards/game-logic.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+    }
+
+    check() {
+      const cardId = this.cards[0].dataset.cardId;
+      const isEqual = this.cards.every(
+        (card) => card.dataset.cardId === cardId
+      );
+      if (isEqual) {
+        winGame += 1;
+        console.log("winGame");
```
Comment (A-Ostrovnyy) [inline submissions/semenjob/memoryCards/game-logic.js:92]: Remove console.log

File: submissions/semenjob/memoryCards/game-logic.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    const selectedCard = target.closest(".memory_card");
+    if (!selectedCard) return;
+    toggleFlip(selectedCard);
+    pair.addCard(selectedCard);
+    if (pair.isFull) {
+      cards.removeEventListener("click", processClickOnCard);
+      setTimeout(() => {
+        cards.addEventListener("click", processClickOnCard);
+      }, 500);
+      const isEquel = pair.check();
```
Comment (A-Ostrovnyy) [inline submissions/semenjob/memoryCards/game-logic.js:141]: Looks like it's typo)

File: submissions/semenjob/memoryCards/game-logic.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+    const card = document.createElement("div");
+    card.classList.add("memory_card", "block");
+    card.innerHTML = flipperTemplate;
+    card.dataset.cardId = id;
+    return card;
+  };
+
+  const cards = document.querySelector(".memory_cards");
+  const cardsLength = 6;
```
Comment (A-Ostrovnyy) [inline submissions/semenjob/memoryCards/game-logic.js:44]: Is this a magic number?)

File: submissions/semenjob/memoryCards/game-logic.js [hunk 7]
```diff
@@ -0,0 +1,152 @@
+"use strict";
+import { memoryCardImages, memoryCardFrontImage } from "./constant.js";
+
+document.addEventListener("DOMContentLoaded", (event) => {
+  startGame();
+});
```
Comment (A-Ostrovnyy) [inline submissions/semenjob/memoryCards/game-logic.js:6]:
```suggestion
document.addEventListener("DOMContentLoaded", startGame);
```

File: submissions/semenjob/memoryCards/game-logic.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+
+    check() {
+      const cardId = this.cards[0].dataset.cardId;
+      const isEqual = this.cards.every(
+        (card) => card.dataset.cardId === cardId
+      );
+      if (isEqual) {
+        winGame += 1;
+        console.log("winGame");
+        addClassBlock(this.cards);
```
Comment (A-Ostrovnyy) [inline submissions/semenjob/memoryCards/game-logic.js:93]: You use function outside your class. One of the main principles of OOP is Encapsulation: all methods of the class must be inside this class.

**Review by A-Ostrovnyy — CHANGES_REQUESTED** [review]
Good work!

**Pull request closed without merging**